File: WildcardMatching.py
class Solution:
    # @param s, an input string
    # @param p, a pattern string
    # @return a boolean
    # http://www.cnblogs.com/zuoyuan/p/3781872.html
    def isMatch(self, s, p):
        """
        @param s_pointer: s pointer
        @param p_pointer: p pointer
        """
        s_pointer = p_pointer = 0
        s_match = 0
        p_star = -1
        while s_pointer < len(s):
            # Match single char or ?
            if p_pointer < len(p) and (s[s_pointer] == p[p_pointer] or p[p_pointer] == '?'):
                s_pointer += 1
                p_pointer += 1
                continue
            # Meet star, save position
            if p_pointer < len(p) and p[p_pointer] == '*':
                p_star = p_pointer
                p_pointer += 1
                s_match = s_pointer
                continue
            if p_star != -1:
                p_pointer = p_star + 1
                s_match += 1
                s_pointer = s_match
                continue
            return False
        while p_pointer < len(p) and p[p_pointer] == '*':
            p_pointer += 1
        if p_pointer ==len(p):
            return True
        return False

if __name__ == '__main__':
    so = Solution()
    print so.isMatch('aaaabaaab', 'a*b*ab')

File: CPP/Second/wildcard_matching.cc
// https://oj.leetcode.com/problems/wildcard-matching/

// Implement wildcard pattern matching with support for '?' and '*'.
// '?' Matches any single character.
// '*' Matches any sequence of characters (including the empty sequence).
// The matching should cover the entire input string (not partial).
// The function prototype should be:
// bool isMatch(const char *s, const char *p)
// Some examples:
// isMatch("aa","a") → false
// isMatch("aa","aa") → true
// isMatch("aaa","aa") → false
// isMatch("aa", "*") → true
// isMatch("aa", "a*") → true
// isMatch("ab", "?*") → true
// isMatch("aab", "c*a*b") → false

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    // http://www.cnblogs.com/zuoyuan/p/3781872.html
    bool isMatch(const char *s, const char *p) {
        // 常量指针， 指向常量的指针
        const char *star = NULL, *ss = s;
        while(*s) {
            // match current position, exactly like '.'
            // also means star match 0
            if (*p == *s || *p == '?')
                p++, s++;
            // hey, there is a star, save it
            else if (*p == '*')
                star = p++, ss = s;
            else if (star)
                p = star+1, s = ++ss;
            else
                return false;
        }
        // let the redundant '*' match zero things
        while (*p == '*')    p++;
        // if p reach the end, true, otherwise, false
        return (!*p);
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
